### PR TITLE
PERF: Fixes performance regression in DataFrame[bool_indexer] (#33924)

### LINKED ISSUE #33924: Performance regression in DataFrame[bool_indexer]
```python
import numpy as np
import pandas as pd

idx_dupe = np.array(range(30)) * 99
df = pd.DataFrame(np.random.randn(10000, 5))
bool_indexer = [True] * 5000 + [False] * 5000


%timeit df[np.array(bool_indexer)]
```

```
# 1.0.2
2.58 ms ± 108 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

# master
5.47 ms ± 192 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

Note that this only affects the case where the indexer is a *list*. Performance looks fine when an ndarray is passed.

https://pandas.pydata.org/speed/pandas/index.html#indexing.DataFrameNumericIndexing.time_bool_indexer?commits=80d37adcc3d9bfbbe17e8aa626d6b5873465ca98-4f89c261f624305fc7bae6c43ae862663994be34 points to somewhere in 80d37adc..4f89c261, which has a bunch of commits.

### PR DESCRIPTION
**Pull request by mproszewska**
- [x] closes #33924
- [x] tests passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Insted of calling `construction.array` in `check_array_indexer`, creates array with `dtype=bool` before calling `check_array_indexer`.

Fixes performance regression after commit b9bcdc307.

```
setup = """
import numpy as np
import pandas as pd
df = pd.DataFrame(np.random.randn(100000, 5))
bool_indexer = [True] * 50000 + [False] * 50000
"""
import timeit
timeit.timeit("df[bool_indexer]",setup=setup, number=1000)

# master
# 27.29123323399108
# now
# 8.814320757053792
```

### CODE REVIEW

File: pandas/core/indexing.py [hunk 1]
```diff
@@ -2164,11 +2166,14 @@ def check_bool_indexer(index: Index, key) -> np.ndarray:
         result = result.astype(bool)._values
     elif is_object_dtype(key):
         # key might be object-dtype bool, check_array_indexer needs bool array
+
         result = np.asarray(result, dtype=bool)
         result = check_array_indexer(index, result)
     else:
+        if not is_array_like(result):
+            # GH 33924
+            result = pd_array(result, dtype=bool)
```
Comment (jbrockmendel) [inline pandas/core/indexing.py:2175]: Why not just np.array?

Comment (mproszewska) [reply pandas/core/indexing.py:2175]: because of nullable input

Comment (jbrockmendel) [reply pandas/core/indexing.py:2174]: if the input is nullable, wouldn't `not is_array_like(result)` be `False`?

Comment (mproszewska) [reply pandas/core/indexing.py:2174]: I meant that this array/list can have nullable elements.

**Review by jreback — CHANGES_REQUESTED** [review]
can you add an asv that hits this case & a whatsnew note

File: pandas/core/indexing.py [hunk 2]
```diff
@@ -2167,8 +2169,10 @@ def check_bool_indexer(index: Index, key) -> np.ndarray:
         result = np.asarray(result, dtype=bool)
         result = check_array_indexer(index, result)
     else:
```
Comment (jreback) [inline pandas/core/indexing.py:2171]: can you make this an elif, then have a separate else clause.

put comments on what each is catching

**Comment by mproszewska** [comment]
`indexing.DataFrameNumericIndexing.time_bool_indexer` looks smiliar to the case that linked issue describes, so I'm not sure if additional asv is needed

File: doc/source/whatsnew/v1.1.0.rst [hunk 3]
```diff
@@ -616,6 +616,7 @@ Performance improvements
 - Performance improvement in :func:`factorize` for nullable (integer and boolean) dtypes (:issue:`33064`).
 - Performance improvement in reductions (sum, prod, min, max) for nullable (integer and boolean) dtypes (:issue:`30982`, :issue:`33261`, :issue:`33442`).
 - Performance improvement in arithmetic operations between two :class:`DataFrame` objects (:issue:`32779`)
+- Performance improvement in `indexing.check_bool_indexer` when `key` is a list (:issue:`33924`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.1.0.rst:619]: this is not user facing, can you rewrite

File: pandas/core/indexing.py [hunk 4]
```diff
@@ -2166,9 +2168,13 @@ def check_bool_indexer(index: Index, key) -> np.ndarray:
         # key might be object-dtype bool, check_array_indexer needs bool array
         result = np.asarray(result, dtype=bool)
         result = check_array_indexer(index, result)
+    elif not is_array_like(result):
+        # GH 33924
+        # key may contain nan elements, check_array_indexer needs bool array
+        result = pd_array(result, dtype=bool)
```
Comment (jreback) [inline pandas/core/indexing.py:2174]: can you make this a bit simpler here, pls do this:

L2166, change that to: `return result.astype(bool)._values`
L21067: elif -> if
rm L2170
rm L2174
rm L2176-2177

L2178: return check_array_indexer(indexer, result)

basically pulling the last 3 clauses together

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
can you run and post the indexing the asvs

**Comment by mproszewska** [comment]
```
[ 50.00%] · For pandas commit e0a34f8f <perf-bool> (round 2/2):
[ 50.00%] ·· Benchmarking conda-py3.6-Cython0.29.16-matplotlib-numba-numexpr-numpy-odfpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[ 55.00%] ··· indexing.DataFrameNumericIndexing.time_bool_indexer                                                                                    1.29±0.3ms
[ 60.00%] ··· indexing.DataFrameNumericIndexing.time_iloc                                                                                              147±20μs
[ 65.00%] ··· indexing.DataFrameNumericIndexing.time_iloc_dups                                                                                         204±20μs
[ 70.00%] ··· indexing.DataFrameNumericIndexing.time_loc                                                                                               83.2±4μs
[ 75.00%] ··· indexing.DataFrameNumericIndexing.time_loc_dups                                                                                        4.73±0.3ms
[ 75.00%] · For pandas commit 5f26c342 <master^2> (round 2/2):
[ 75.00%] ·· Building for conda-py3.6-Cython0.29.16-matplotlib-numba-numexpr-numpy-odfpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt...
[ 75.00%] ·· Benchmarking conda-py3.6-Cython0.29.16-matplotlib-numba-numexpr-numpy-odfpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[ 80.00%] ··· indexing.DataFrameNumericIndexing.time_bool_indexer                                                                                    3.63±0.1ms
[ 85.00%] ··· indexing.DataFrameNumericIndexing.time_iloc                                                                                               116±6μs
[ 90.00%] ··· indexing.DataFrameNumericIndexing.time_iloc_dups                                                                                         180±20μs
[ 95.00%] ··· indexing.DataFrameNumericIndexing.time_loc                                                                                               85.1±9μs
[100.00%] ··· indexing.DataFrameNumericIndexing.time_loc_dups                                                                                        5.86±0.8ms
```

**Comment by mproszewska** [comment]
I could change size of tested DataFrame, so that benchmarks would change more significantly.

**Comment by jreback** [comment]
> I could change size of tested DataFrame, so that benchmarks would change more significantly.

sure as long as benchamarks are << 1s its ok (e.g. in ms). please post the actual asv run. otherwise looks fine. ping on green.

**Comment by mproszewska** [comment]
After asv change
```
 before           after         ratio
     [5f26c342]       [e0a34f8f]
     <master^2>       <perf-bool>
-      26.3±0.3ms      7.27±0.03ms     0.28  indexing.DataFrameNumericIndexing.time_bool_indexer
```

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm ping on green

**Comment by mproszewska** [comment]
> lgtm ping on green

ping

**Comment by jreback** [comment]
thanks @mproszewska

**Pull request merged**